### PR TITLE
[AN-4387] Adds ability to edit text based messages

### PR DESCRIPTION
**Pull request by shumengye**
#### Description
Adds ability to edit text based messages

**Note!**
Do not merge, SE implementation for edit not ready yet

#### Ticket
https://wearezeta.atlassian.net/browse/AN-4387
#### APK

[Download build #7360](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7360/artifact/build/artifact/wire-dev-PR42-7360.apk)
[Download build #7361](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7361/artifact/build/artifact/wire-dev-PR42-7361.apk)

### CODE REVIEW

**Comment by shumengye** [comment]
@mykola-mokhnach 
Re-created PR for https://github.com/wireapp/wire-android/pull/39

Next build should have fix for:
- Dark theme issues
- Back press closed edit
- Only messages from myself should be editable

**Comment by shumengye** [comment]
@mykola-mokhnach 
I'm not able to reproduce the rotation crash on tablet. Can you check if it still happens, if yes on which device?

**Comment by mykola-mokhnach** [comment]
Edit option should not be available for own messages in a group conversation after you leave this conversation or if somebody else removed you from this convo. confirmed with designers

**Comment by mykola-mokhnach** [comment]
@shumengye I cannot repro the issue with tablet as well. It seems to be random. but please check the traceback - maybe we miss some NPE verification in code.

**Comment by mykola-mokhnach** [comment]
It might be confusing for user if he edits a message from history, because the view is always automatically scrolled to the very bottom after a message edit control is shown. But it's more a question to designers.

**Comment by mykola-mokhnach** [comment]
Clicking Edit on Soundcloud preview has no effect

**Comment by shumengye** [comment]
@mykola-mokhnach Edit should work only on text and link preview messages

**Comment by mykola-mokhnach** [comment]
Same problem with youtube and spotify previews

**Comment by shumengye** [comment]
@mykola-mokhnach Next build won't show EDIT option for SoundCloud, YouTube etc. rich media messages

File: app/src/main/java/com/waz/zclient/pages/main/conversation/ConversationFragment.java [hunk 1]
```diff
@@ -868,7 +873,8 @@ public void run() {
 
 
                 // handle draft
-                if (fromConversation != null && changeToDifferentConversation) {
+                if (fromConversation != null && changeToDifferentConversation &&
+                    !cursorLayout.isEditingMessage()) {
```
Comment (hakonbo) [inline app/src/main/java/com/waz/zclient/pages/main/conversation/ConversationFragment.java]: I guess if we change to a different conversation, we should stop message editing and show any potential draft message for the new conversation?

Comment (shumengye) [inline app/src/main/java/com/waz/zclient/pages/main/conversation/ConversationFragment.java]: The check is for saving the draft. Text in the edit text field should not be saved as a draft for the conversation.

**Comment by mykola-mokhnach** [comment]
1. Open a group conversation with userB and userC
2. Long tap your message in this conversation and select Edit
3. Remove yourself as user B from this conversation

AR:
'Edit message' control is not visible anymore, but keyboard is still there

ER:
I expect the keyboard to be automatically hidden after STR

**Comment by mykola-mokhnach** [comment]
It would be nice not to scroll the conversation to the very bottom if I select Edit Message option. This is confusing for a user, which does not see the result of message editing immediately, but is forced to scroll up instead.

**Comment by shumengye** [comment]
@mykola-mokhnach 
Next build has changes
- Edit and delete everywhere are not possible if user is not member of conversation
- Edit is allowed for all rich media type messages

I'll investigate the other issues you posted

**Comment by xuanzhaopeng** [comment]
@shumengye  we have test : 

Verify I can cancel editing a message by tap on something else

Currently it doesn't work

**Comment by shumengye** [comment]
@mykola-mokhnach 
Next build 
- won't scroll conversation to the bottom when editing message
- integrated SE version with edit capacity

File: wire-core/src/main/res/values/dimens.xml [hunk 2]
```diff
@@ -543,7 +543,7 @@
     <dimen name="drawing__bottom_bar__height">80dp</dimen>
     <dimen name="drawing__button__padding">8dp</dimen>
     <dimen name="drawing__send__button__padding">10dp</dimen>
-    <dimen name="drawing__button_send__text_size">40dp</dimen>
+    <dimen name="drawing__button_send__text_size">24dp</dimen>
```
Comment (hakonbo) [inline wire-core/src/main/res/values/dimens.xml]: Is this change part of this PR intentionally?

File: wire-ui/src/main/res/layout/edit_message_cursor_toolbar.xml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+             android:layout_height="match_parent"
+             android:layout_marginLeft="@dimen/wire__padding__regular"
+             android:layout_marginRight="@dimen/wire__padding__regular">
+
+    <com.waz.zclient.ui.text.GlyphTextView
+        android:id="@+id/gtv__edit_message__reset"
+        android:layout_width="wrap_content"
+        android:layout_height="match_parent"
+        android:layout_gravity="left|center_vertical"
+        android:padding="@dimen/drawing__button__padding"
```
Comment (hakonbo) [inline wire-ui/src/main/res/layout/edit_message_cursor_toolbar.xml]: All the ```@dimen/drawing__``` refs are for the drawing fragment. Maybe best to create new ones here, so we don't accidentally change both when only wanting to change one in the future.

File: wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java [hunk 4]
```diff
@@ -425,6 +458,70 @@ public void run() {
...
+        if (message == null) {
+            return;
+        }
+        newCursorEditText.setText(message.getBody());
+        newCursorEditText.setSelection(newCursorEditText.getText().length());
+    }
+
+    @Override
+    public void onApproveEditMessage() {
+        message.update(new MessageContent.Text(newCursorEditText.getText().toString()));
```
Comment (hakonbo) [inline wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java]: Maybe we want a check for 

1. if there is actually a change
2. if the new edit text is empty (then maybe we can just delete the message instead?)

File: wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java [hunk 5]
```diff
@@ -425,6 +458,70 @@ public void run() {
...
+        message = null;
+        newCursorEditText.setText("");
+        hintView.setVisibility(VISIBLE);
+        isEditingMessage = false;
+
+        if (animated) {
+            hideEditMessageToolbar();
+            ViewUtils.fadeOutView(editMessageBackgroundView);
+        }
+        else {
```
Comment (hakonbo) [inline wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java]: ```else``` should be on previous line

File: wire-core/src/main/res/values/dimens.xml (hunk 2 above)
Comment (shumengye) [inline wire-core/src/main/res/values/dimens.xml]: yes, the checkmark is too big here and in sketch (requested by design)

File: wire-ui/src/main/res/layout/edit_message_cursor_toolbar.xml (hunk 3 above)
Comment (shumengye) [inline wire-ui/src/main/res/layout/edit_message_cursor_toolbar.xml]: ok will do

File: wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java (hunk 4 above)
Comment (shumengye) [inline wire-ui/src/main/java/com/waz/zclient/ui/cursor/CursorLayout.java]: We have a check in the UI - the submit button is disabled if there was no change.

**Comment by mykola-mokhnach** [comment]
It should not be possible to commit empty message/message containing non-printable characters (\\s+) after edit

**Comment by mykola-mokhnach** [comment]
Pen icon does not appear on the other side after a message is edited

**Comment by shumengye** [comment]
@mykola-mokhnach 
Did you test for other side with #7361?

**Comment by mykola-mokhnach** [comment]
@shumengye I have just started it.
One issue has been found so far

> Pen icon does not appear on the other side after a message is edited

**Comment by hakonbo** [comment]
@mykola-mokhnach @shumengye will merge this PR now as @Yorgos-V needs to start testing. Will be available in the Experimental build.

**Pull request merged**